File: SyntheticData+AI/frontend/src/lib/offline/features.ts
/**
 * The 26 diagnostic features, in the exact order the model expects.
 *
 * A faithful port of `vfactory/features.py`. Order matters: the weight bundle
 * indexes features positionally, so a reordering here would silently produce
 * nonsense rather than an error. The parity script guards against that.
 */

import {
  amplitudeSpectrum,
  envelopeSpectrum,
  mean,
  median,
  velocityRmsMmS,
  type Spectrum,
} from "./dsp.ts";

const EPS = 1e-12;

export const SPECTRAL_BANDS: Array<[number, number]> = [
  [0, 200],
  [200, 600],
  [600, 1500],
  [1500, 2500],
  [2500, 3500],
  [3500, 6000],
];

export const ENVELOPE_BAND: [number, number] = [2000, 4500];

export const FEATURE_NAMES = [
  "rms",
  "peak",
  "peak_to_peak",
  "crest_factor",
  "kurtosis",
  "skewness",
  "shape_factor",
  "impulse_factor",
  "clearance_factor",
  "zero_crossing_rate",
  "band_0_200_db",
  "band_200_600_db",
  "band_600_1500_db",
  "band_1500_2500_db",
  "band_2500_3500_db",
  "band_3500_6000_db",
  "spectral_centroid",
  "spectral_spread",
  "spectral_entropy",
  "env_shaft_1x_db",
  "env_bpfo_db",
  "env_bpfo_h2_db",
  "env_bpfi_db",
  "env_bpfi_h2_db",
  "env_bsf_db",
  "env_ftf_db",
] as const;

export type FeatureName = (typeof FEATURE_NAMES)[number];

/** Envelope probes as [defect key, harmonic], matching ENVELOPE_PROBES. */
const ENVELOPE_PROBES: Array<[string, number]> = [
  ["shaft", 1],
  ["bpfo", 1],
  ["bpfo", 2],
  ["bpfi", 1],
  ["bpfi", 2],
  ["bsf", 1],
  ["ftf", 1],
];

export interface BearingOrders {
  bpfo: number;
  bpfi: number;
  bsf: number;
  ftf: number;
}

export interface MachineSpec {
  sampleRateHz: number;
  windowSize: number;
  shaftRpm: number;
  resonanceHz: number;
  resonanceDecay: number;
  noiseG: number;
  baselineImbalanceG: number;
  orders: BearingOrders;
}

export const DEFAULT_MACHINE: MachineSpec = {
  sampleRateHz: 12000,
  windowSize: 2048,
  shaftRpm: 1797,
  resonanceHz: 3000,
  resonanceDecay: 900,
  noiseG: 0.035,
  baselineImbalanceG: 0.035,
  // SKF 6205-2RS geometry, as multiples of shaft speed.
  orders: { bpfo: 3.5848, bpfi: 5.4152, bsf: 2.3567, ftf: 0.3983 },
};

/** Matches `np.signbit`: negative zero counts as negative. */
function signbit(value: number): boolean {
  return value < 0 || Object.is(value, -0);
}

function timeFeatures(signal: Float64Array): number[] {
  const n = signal.length;
  const avg = mean(signal);

  let sumSquares = 0;
  let peak = 0;
  let min = Infinity;
  let max = -Infinity;
  let absMean = 0;
  let sqrtMean = 0;
  let m2 = 0;
  let m3 = 0;
  let m4 = 0;
  let crossings = 0;

  for (let i = 0; i < n; i++) {
    const v = signal[i];
    sumSquares += v * v;
    const abs = Math.abs(v);
    if (abs > peak) peak = abs;
    if (v < min) min = v;
    if (v > max) max = v;
    absMean += abs;
    sqrtMean += Math.sqrt(abs);
    const c = v - avg;
    m2 += c * c;
    m3 += c * c * c;
    m4 += c * c * c * c;
    if (i > 0 && signbit(v) !== signbit(signal[i - 1])) crossings++;
  }

  const rms = Math.sqrt(sumSquares / n);
  absMean = absMean / n + EPS;
  sqrtMean = (sqrtMean / n) ** 2 + EPS;
  const std = Math.sqrt(m2 / n) + EPS;

  return [
    rms,
    peak,
    max - min,
    peak / (rms + EPS),
    m4 / n / std ** 4,
    m3 / n / std ** 3,
    rms / absMean,
    peak / absMean,
    peak / sqrtMean,
    crossings / (n - 1),
  ];
}

function spectralFeatures({ freqs, amp }: Spectrum): number[] {
  const bins = freqs.length;
  const power = new Float64Array(bins);
  let total = 0;
  for (let k = 0; k < bins; k++) {
    power[k] = amp[k] * amp[k];
    total += power[k];
  }
  total += EPS;

  const bands = SPECTRAL_BANDS.map(([lo, hi]) => {
    let sum = 0;
    for (let k = 0; k < bins; k++) {
      if (freqs[k] >= lo && freqs[k] < hi) sum += power[k];
    }
    return 10 * Math.log10(sum / total + EPS);
  });

  let centroid = 0;
  for (let k = 0; k < bins; k++) centroid += freqs[k] * power[k];
  centroid /= total;

  let spread = 0;
  let entropy = 0;
  for (let k = 0; k < bins; k++) {
    spread += (freqs[k] - centroid) ** 2 * power[k];
    const p = power[k] / total;
    entropy += p * Math.log2(p + EPS);
  }
  spread = Math.sqrt(spread / total);
  entropy = -entropy / Math.log2(bins);

  return [...bands, centroid, spread, entropy];
}

function envelopeFeatures(
  { freqs, amp }: Spectrum,
  defectHz: Record<string, number>,
  toleranceHz: number,
): number[] {
  const floor = median(amp) + EPS;
  return ENVELOPE_PROBES.map(([key, harmonic]) => {
    const target = defectHz[key] * harmonic;
    let peak = 0;
    for (let k = 0; k < freqs.length; k++) {
      if (Math.abs(freqs[k] - target) <= toleranceHz && amp[k] > peak) peak = amp[k];
    }
    return 20 * Math.log10(peak / floor + EPS);
  });
}

export function defectFrequencies(
  machine: MachineSpec,
  shaftRpm: number,
): Record<string, number> {
  const shaftHz = shaftRpm / 60;
  return {
    shaft: shaftHz,
    bpfo: machine.orders.bpfo * shaftHz,
    bpfi: machine.orders.bpfi * shaftHz,
    bsf: machine.orders.bsf * shaftHz,
    ftf: machine.orders.ftf * shaftHz,
  };
}

export interface Extraction {
  values: Float64Array;
  spectrum: Spectrum;
  envelope: Spectrum;
  velocityRmsMmS: number;
}

/** Condense one acceleration window into the diagnostic feature vector. */
export function extract(
  signal: Float64Array,
  machine: MachineSpec = DEFAULT_MACHINE,
  shaftRpm?: number,
): Extraction {
  const rpm = shaftRpm ?? machine.shaftRpm;
  const spectrum = amplitudeSpectrum(signal, machine.sampleRateHz);
  const envelope = envelopeSpectrum(signal, machine.sampleRateHz, ENVELOPE_BAND);
  // Slip and speed drift smear the defect lines; allow two FFT bins either way.
  const tolerance = (2 * machine.sampleRateHz) / signal.length;

  const values = Float64Array.from([
    ...timeFeatures(signal),
    ...spectralFeatures(spectrum),
    ...envelopeFeatures(envelope, defectFrequencies(machine, rpm), tolerance),
  ]);

  return {
    values,
    spectrum,
    envelope,
    velocityRmsMmS: velocityRmsMmS(signal, machine.sampleRateHz),
  };
}

export function describe(values: Float64Array): Record<string, number> {
  const out: Record<string, number> = {};
  FEATURE_NAMES.forEach((name, i) => {
    out[name] = values[i];
  });
  return out;
}

File: SyntheticData+AI/src/vfactory/features.py
"""Feature extraction: turn a raw acceleration window into a diagnostic vector.

An autoencoder over a *single scalar* sample (what the original project did)
cannot learn anything -- a 1-to-4-to-1 network trivially learns the identity
and its reconstruction error carries no information about machine health.
The signal only becomes separable once each window is condensed into the
features a vibration analyst would actually look at:

* **time domain** -- RMS, peak, crest, kurtosis and the other shape factors
  that react to impulsiveness;
* **frequency domain** -- band energies, spectral centroid/spread/entropy;
* **envelope domain** -- amplitude demodulation around the structural
  resonance, read at the bearing's geometric defect frequencies. This is the
  standard way a localised bearing defect is detected, and it is what lets the
  model distinguish an outer-race spall from plain imbalance.

Everything is NumPy-only so the same code can run inside a request handler
without dragging a training stack into production.
"""

from __future__ import annotations

import numpy as np

from .config import (
    DEFAULT_MACHINE,
    ENVELOPE_BAND_HZ,
    SPECTRAL_BANDS,
    MachineSpec,
)

_EPS = 1e-12

#: Equivalent noise bandwidth of a Hann window, in bins. A tapered window
#: smears a tone across about three bins, so summing power across a band
#: over-counts it by this factor unless it is divided back out. Amplitude
#: readings of a single peak are unaffected -- only band power needs it.
_HANN_ENBW = 1.5

TIME_FEATURES: tuple[str, ...] = (
    "rms",
    "peak",
    "peak_to_peak",
    "crest_factor",
    "kurtosis",
    "skewness",
    "shape_factor",
    "impulse_factor",
    "clearance_factor",
    "zero_crossing_rate",
)

SPECTRAL_FEATURES: tuple[str, ...] = (
    *(f"band_{int(lo)}_{int(hi)}_db" for lo, hi in SPECTRAL_BANDS),
    "spectral_centroid",
    "spectral_spread",
    "spectral_entropy",
)

#: Envelope-spectrum probes, as (feature name, defect key, harmonic).
ENVELOPE_PROBES: tuple[tuple[str, str, int], ...] = (
    ("env_shaft_1x_db", "shaft", 1),
    ("env_bpfo_db", "bpfo", 1),
    ("env_bpfo_h2_db", "bpfo", 2),
    ("env_bpfi_db", "bpfi", 1),
    ("env_bpfi_h2_db", "bpfi", 2),
    ("env_bsf_db", "bsf", 1),
    ("env_ftf_db", "ftf", 1),
)

ENVELOPE_FEATURES: tuple[str, ...] = tuple(name for name, _, _ in ENVELOPE_PROBES)

FEATURE_NAMES: tuple[str, ...] = TIME_FEATURES + SPECTRAL_FEATURES + ENVELOPE_FEATURES
N_FEATURES = len(FEATURE_NAMES)


# --------------------------------------------------------------------------- #
# Spectral helpers (also used by the API to draw charts)
# --------------------------------------------------------------------------- #


def amplitude_spectrum(
    window: np.ndarray, sample_rate_hz: float
) -> tuple[np.ndarray, np.ndarray]:
    """Single-sided amplitude spectrum of a Hann-windowed signal."""
    n = window.shape[-1]
    taper = np.hanning(n)
    # Coherent gain of a Hann window is 0.5; correct so peak heights are right.
    scaled = (window - window.mean(axis=-1, keepdims=True)) * taper
    spec = np.fft.rfft(scaled, axis=-1)
    amp = 2.0 * np.abs(spec) / (n * 0.5)
    freqs = np.fft.rfftfreq(n, 1.0 / sample_rate_hz)
    return freqs, amp


def envelope(window: np.ndarray, sample_rate_hz: float, band_hz: tuple[float, float]) -> np.ndarray:
    """Hilbert envelope of the signal band-passed around the resonance.

    Band-passing first is what makes envelope analysis work: the resonance
    carries the impact energy at a high signal-to-noise ratio, while the large
    low-frequency shaft components that would otherwise dominate are removed.
    """
    n = window.shape[-1]
    spec = np.fft.fft(window - window.mean(axis=-1, keepdims=True), axis=-1)
    freqs = np.fft.fftfreq(n, 1.0 / sample_rate_hz)

    lo, hi = band_hz
    keep = (np.abs(freqs) >= lo) & (np.abs(freqs) <= hi)
    spec = spec * keep

    # Analytic signal: zero the negative frequencies, double the positive ones.
    analytic_gain = np.zeros(n)
    analytic_gain[0] = 1.0
    if n % 2 == 0:
        analytic_gain[n // 2] = 1.0
        analytic_gain[1 : n // 2] = 2.0
    else:
        analytic_gain[1 : (n + 1) // 2] = 2.0

    return np.abs(np.fft.ifft(spec * analytic_gain, axis=-1))


def envelope_spectrum(
    window: np.ndarray,
    sample_rate_hz: float,
    band_hz: tuple[float, float] = ENVELOPE_BAND_HZ,
) -> tuple[np.ndarray, np.ndarray]:
    """Spectrum of the envelope -- where bearing defect lines appear."""
    env = envelope(window, sample_rate_hz, band_hz)
    return amplitude_spectrum(env, sample_rate_hz)


def velocity_rms_mm_s(
    window: np.ndarray, sample_rate_hz: float, band_hz: tuple[float, float] = (10.0, 1_000.0)
) -> float:
    """RMS velocity over the ISO 20816 band, converted from acceleration in g."""
    freqs, amp = amplitude_spectrum(window, sample_rate_hz)
    lo, hi = band_hz
    keep = (freqs >= lo) & (freqs <= hi)
    if not keep.any():
        return 0.0
    # a[g] -> a[m/s^2] -> v[m/s] by dividing by omega -> v[mm/s].
    vel_amp = amp[keep] * 9.80665 * 1000.0 / (2.0 * np.pi * freqs[keep])
    power = np.sum((vel_amp / np.sqrt(2.0)) ** 2) / _HANN_ENBW
    return float(np.sqrt(power))


def _peak_near(
    freqs: np.ndarray, values: np.ndarray, target_hz: float, tolerance_hz: float
) -> float:
    """Largest value within +/- tolerance of a target frequency."""
    keep = np.abs(freqs - target_hz) <= tolerance_hz
    return float(values[keep].max()) if keep.any() else 0.0


# --------------------------------------------------------------------------- #
# Feature extraction
# --------------------------------------------------------------------------- #


def _time_features(window: np.ndarray) -> list[float]:
    centred = window - window.mean()
    rms = float(np.sqrt(np.mean(window**2)))
    peak = float(np.max(np.abs(window)))
    p2p = float(window.max() - window.min())
    abs_mean = float(np.mean(np.abs(window))) + _EPS
    sqrt_mean = float(np.mean(np.sqrt(np.abs(window)))) ** 2 + _EPS
    std = float(centred.std()) + _EPS

    return [
        rms,
        peak,
        p2p,
        peak / (rms + _EPS),                       # crest factor
        float(np.mean(centred**4) / std**4),       # kurtosis
        float(np.mean(centred**3) / std**3),       # skewness
        rms / abs_mean,                            # shape factor
        peak / abs_mean,                           # impulse factor
        peak / sqrt_mean,                          # clearance factor
        float(np.mean(np.abs(np.diff(np.signbit(window))))),  # zero-crossing rate
    ]


def _spectral_features(freqs: np.ndarray, amp: np.ndarray) -> list[float]:
    power = amp**2
    total = float(power.sum()) + _EPS

    band_energies = []
    for lo, hi in SPECTRAL_BANDS:
        keep = (freqs >= lo) & (freqs < hi)
        band_energies.append(10.0 * np.log10(float(power[keep].sum()) / total + _EPS))

    centroid = float((freqs * power).sum() / total)
    spread = float(np.sqrt(((freqs - centroid) ** 2 * power).sum() / total))
    p = power / total
    entropy = float(-(p * np.log2(p + _EPS)).sum() / np.log2(len(p)))

    return [*band_energies, centroid, spread, entropy]


def _envelope_features(
    freqs: np.ndarray, amp: np.ndarray, defect_hz: dict[str, float], tolerance_hz: float
) -> list[float]:
    """Prominence of each defect line above the envelope noise floor, in dB.

    Using a ratio against the local median rather than a raw amplitude keeps
    the feature meaningful when the overall vibration level changes -- which is
    what separates "a defect line grew" from "the whole machine got louder".
    """
    floor = float(np.median(amp)) + _EPS
    out = []
    for _, key, harmonic in ENVELOPE_PROBES:
        target = defect_hz[key] * harmonic
        peak = _peak_near(freqs, amp, target, tolerance_hz)
        out.append(20.0 * np.log10(peak / floor + _EPS))
    return out


def extract(
    window: np.ndarray,
    machine: MachineSpec = DEFAULT_MACHINE,
    shaft_rpm: float | None = None,
) -> np.ndarray:
    """Condense one acceleration window into the diagnostic feature vector."""
    window = np.asarray(window, dtype=np.float64).ravel()
    rpm = shaft_rpm if shaft_rpm else machine.shaft_rpm

    freqs, amp = amplitude_spectrum(window, machine.sample_rate_hz)
    env_freqs, env_amp = envelope_spectrum(window, machine.sample_rate_hz)

    defect_hz = machine.bearing.defect_frequencies(rpm)
    defect_hz["shaft"] = rpm / 60.0
    # Slip and speed drift smear the defect lines; allow two FFT bins either way.
    tolerance = 2.0 * machine.sample_rate_hz / window.size

    values = (
        _time_features(window)
        + _spectral_features(freqs, amp)
        + _envelope_features(env_freqs, env_amp, defect_hz, tolerance)
    )
    return np.asarray(values, dtype=np.float64)


def extract_batch(
    windows: np.ndarray,
    machine: MachineSpec = DEFAULT_MACHINE,
    shaft_rpm: float | None = None,
) -> np.ndarray:
    """Feature matrix for a stack of windows, shape ``(n_windows, N_FEATURES)``."""
    return np.stack([extract(w, machine, shaft_rpm) for w in np.atleast_2d(windows)])


def describe(values: np.ndarray) -> dict[str, float]:
    """Pair a feature vector with its names, for logging and API responses."""
    return {name: float(v) for name, v in zip(FEATURE_NAMES, values, strict=True)}
